caddyhttp/staticfiles/fileserver.go

```diff
@@ -80,17 +80,35 @@ func (fs FileServer) serveFile(w http.ResponseWriter, r *http.Request, name stri
 	}
 
 	// redirect to canonical path
-	url := r.URL.Path
+	u := r.URL.Path
 	if d.IsDir() {
 		// Ensure / at end of directory url
-		if !strings.HasSuffix(url, "/") {
-			Redirect(w, r, path.Base(url)+"/", http.StatusMovedPermanently)
+		if !strings.HasSuffix(u, "/") {
+			// Since the path in u may have been trimmed by Caddy, and Redirect
+			// requires the full path, parse the request's RequestURI and use the
+			// path.
+			redirurl, err := url.ParseRequestURI(r.RequestURI)
```

Comment:
let's rename this to `redirURL` or anything else that won't give lint/vet error.